.github/workflows/release.yml

```diff
@@ -13,38 +13,46 @@ jobs:
     steps:
     - name: Checkout
       uses: actions/checkout@v2
+      with:
+        fetch-depth: 0
 
     - name: Install Node.js
       uses: actions/setup-node@v1
       with:
         node-version: 14.x
 
-    - name: Configure git
+    - name: Get current version
       run: |
-        git config --local user.email "office@geowerkstatt.ch"
-        git config --local user.name "GeoWerkstatt-Build"
+          VERSION=$(node -p "require('./package.json').version")
+          VERSION_TAG=v$VERSION
+          echo VERSION_TAG=$VERSION_TAG >> $GITHUB_ENV
+
+    - name: Fail on existing version tag
+      id: check_tag
+      continue-on-error: true
+      run: exit $(git tag -l ${{ env.VERSION_TAG }} | wc -l) 
 
     - run: npm ci
+      if: ${{ steps.check_tag.action == 'failure' }}
```

Comment:
Q: Kann man hier den exit code invertieren, damit der anschliessende `if` besser lesbarer wird?

Reply:
Vermutlich würde

    !exit $(git tag -l ${{ env.VERSION_TAG }} | wc -l)

gehen. Oder etwas mit `test` und `eq 1` anstelle von `exit`.

Aber

    git rev-parse --verify -q "refs/tags/${{ env.VERSION_TAG }}"

müsste eigentlich auch das Richtige machen. Bzw. um den Output loszuwerden:

    git rev-parse --verify -q "refs/tags/${{ env.VERSION_TAG }}" >> /dev/null

Reply:
Hier ist mir ein Change verlohren gegangen. Eigentlich sollte da `if: ${{ steps.check_tag.outcome == 'success' }}` stehen. Dann sollte es auch einfach verständlich sein.

Reply:
Braucht es dann `continue-on-error: true` überhaupt noch?

Irgendwas geht hier nicht auf, `exit $(git tag -l ${{ env.VERSION_TAG }} | wc -l)` gibt doch 1 zurück, wenn der Tag existiert, und sonst 0. Und gemäss Konvention ist 0 Success, und alles andere ein Fehler. Oder verstehe ich da etwas falsch?

Reply:
GitHub unterscheidet pro step nach `outcome` und `conclusion` ([Doc](https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#steps-context)). Wir haben hier, mit dem Flag `continue-on-error: true`, 2 Fälle:
 - Tag exisiert -> exit 1 -> `outcome` = `failure` jedoch `conclusion`=`success`
 - Tag ist verfügbar -> exit 0 -> `outcome` wie auch `conclusion`=`success`

Mit dieser Lösung wird der Build nicht als Failed gewertet, wenn ein 'normaler' nicht Versions-PR gemerged wird.

Reply:
Ich hab die Doku auch gelesen. Habs aber nicht verstanden :-D
> The result of a **completed step before continue-on-error is applied**.

Auf was bezieht sich `step`? Der GitHub Actions step "_Get current version_"? Der wäre ja immer erfolgreich...
Oder anders gefratgt. Hast du mal versucht ob das so funktioniert?

Reply:
Den Aspekt nicht-Release Merges als OK zu taxieren hab ich total übersehen.

Aber wäre es dann nicht klarer, auf die ganze Success/Fail-Thematik (und continue on error) zu verzichten und diesen Aspekt im Environment zu speichern, z.B.

    echo PUBLISH_NEW_RELEASE=$(git tag -l ${{ env.VERSION_TAG }} | wc -l) >> $GITHUB_ENV

und dann in den Folge-Steps auf diese Variable zu prüfen?

(Oder separate Jobs zu erstellen, damit man nicht daran denken muss bei jedem Step zu prüfen.)

Reply:
Ich hatte auch überlegt: `release.yml` mit trigger auf Tags mit `v*` sowie ein `create-version-tag.yml` mit trigger auf PR-merges. Für diesen Anwendungsfall gibt es auch [bereits Actions](https://github.com/marketplace/actions/github-tag) welche sehr ähnlich angewendet werden.

Reply:
Das Problem am Trigger auf `v*`-Tags ist, dass es ziemlich leicht ist, solche beim „Bumpen“ per `npm version patch` zu erstellen und aus Versehen zu publizieren. Und wenn man als Reaktion darauf eine Branch-Condition auf `main` und `release/*` (oder so) einführt, wird in diesen Fällen kein Release mehr erstellt.

Es wäre einfacher, wenn es einen einfachen, nicht-Fehler-Weg gäbe, den aktuellen Job/Workflow zu stoppen: Eine Action, analog `exit 0` in einem Shell Script. (Es gibt zwar [cancel a workflow run](https://docs.github.com/en/rest/reference/actions#cancel-a-workflow-run) im Actions-API, was ich aber so zum Thema lese deutet eher darauf hin, dass sich diese Methode Zeit lassen kann bis der Job wirklich gestoppt wird.)

Reply:
Eine vermutlich etwas dumme Frage, aber weshlab machen wir Zeilen 29-57 nicht per Shell-Skript, wo _if tag does not exist, exit w/ success_ einfach wäre?

https://github.com/GeoWerkstatt/vsc_interlis2_extension/blob/949a428ec2f5fbd2f71412f1ebe58e1f4a22dae5/.github/workflows/release.yml#L29-L57